pkg/schema/tracker.go

```diff
@@ -24,28 +24,89 @@ import (
 	"github.com/pingcap/parser/model"
 	"github.com/pingcap/tidb-tools/pkg/dbutil"
 	"github.com/pingcap/tidb-tools/pkg/filter"
+	tidbConfig "github.com/pingcap/tidb/config"
 	"github.com/pingcap/tidb/domain"
 	"github.com/pingcap/tidb/infoschema"
 	"github.com/pingcap/tidb/kv"
 	"github.com/pingcap/tidb/meta"
 	"github.com/pingcap/tidb/session"
 	"github.com/pingcap/tidb/store/mockstore"
+
+	"github.com/pingcap/dm/pkg/conn"
+	tcontext "github.com/pingcap/dm/pkg/context"
 )
 
 const (
 	waitDDLRetryCount = 10
 	schemaLeaseTime   = 10 * time.Millisecond
 )
 
+var (
+	sessionVars = []string{"sql_mode", "tidb_skip_utf8_check"}
+)
+
 // Tracker is used to track schema locally.
 type Tracker struct {
 	store kv.Storage
 	dom   *domain.Domain
 	se    session.Session
 }
 
-// NewTracker creates a new tracker.
-func NewTracker(sessionCfg map[string]string) (*Tracker, error) {
+// NewTracker creates a new tracker. `sessionCfg` will be set as tracker's session variables if specified, or retrieve
+// some variable from downstream TiDB using `tidbConn`. `tidbConn` is also used to get some TiDB config
+func NewTracker(sessionCfg map[string]string, trackerCfg *tidbConfig.Config, tidbConn *conn.BaseConn) (*Tracker, error) {
+	var ignoredColumn interface{}
+
+	if trackerCfg != nil {
+		toSet := tidbConfig.NewConfig()
+		// hard to find which field is specified by user, which field is default value by decoding, so only set `alter-primary-key` for now
+		toSet.AlterPrimaryKey = trackerCfg.AlterPrimaryKey
+		tidbConfig.StoreGlobalConfig(toSet)
+	} else {
+		// fetch downstream config, currently only `alter-primary-key`
+
+		// mysql> select json_extract(@@tidb_config, '$."alter-primary-key"');
+		// +------------------------------------------------------+
+		// | json_extract(@@tidb_config, '$."alter-primary-key"') |
+		// +------------------------------------------------------+
+		// | false                                                |
+		// +------------------------------------------------------+
+		rows, err := tidbConn.QuerySQL(tcontext.Background(), "select json_extract(@@tidb_config, '$.\"alter-primary-key\"');")
+		if err != nil {
+			goto SKIP
+		}
+		if rows.Next() {
+			var value bool
+			if err2 := rows.Scan(&value); err2 != nil {
+				goto SKIP
+			}
+			tidbCfg := tidbConfig.NewConfig()
+			tidbCfg.AlterPrimaryKey = value
+			tidbConfig.StoreGlobalConfig(tidbCfg)
+		}
+		rows.Close()
+	}
+SKIP:
+	if len(sessionCfg) == 0 {
+		sessionCfg = make(map[string]string)
+		for _, k := range sessionVars {
+			rows, err2 := tidbConn.QuerySQL(tcontext.Background(), fmt.Sprintf("show variables like '%s'", k))
```

Comment:
we can add `QueryRowContext` for only getting one row into _baseconn_ **later**.